api/v1alpha3/vspherefailuredomain_types.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// VSphereFailureDomainSpec defines the desired state of VSphereFailureDomain
+type VSphereFailureDomainSpec struct {
+	// name of the Region
+	// if is defined with tags, then should be the tag name, for eg: k8s-region-us-west
+	// if not defined with tags for the case with hostgroup, then it’s a user defined region name
+	Region string `json:"region"`
+
+	// name of the Zone,
+	// if is defined with tags, then should be the tag name, for eg:  k8s-zone-us-west-az1
+	// if mapped to hostgroup, then should be the name of the hostgroup
+	Zone string `json:"zone"`
+
+	// Labels is the Tag category used to configure Region/Zone tags
+	// for FailureDomainDatacenter and FailureDomainComputeCluster
+	// +optional
+	Labels LabelsConfig `json:"labels,omitempty"`
+
+	// The underlying infrastructure for this failure domain
+	// Datacenter as the failure domain
+	Datacenter string `json:"datacenter"`
```

Comment:
@jzhoucliqr - moved this to and compute cluster to a string, the reason behind this is the UX, If a user wants to ensure proper spreading in original design, they would always have to rely on the host group field, but can still change the compute cluster in the other `VSphereFailureDomains` something like:

fd1:
```
HostGroup:
  Datacenter: dc1
  ComputeCluster: c1
  Name: "hg1"
```

fd2:
```
HostGroup:
  Datacenter: dc1
  ComputeCluster: c2
  Name: "hg1"
```

this current spec implies that it's a hostgroup based spreading, but at the end it's more a compute cluster based approach

Reply:
Sounds good. With current design, is my following understanding correct ? 
1. datacenter=dc1,computeCluster=nil,hostGroup=nil        => Datacenter dc1 as failure domain
2. datacenter=dc1,computeCluster=c1,hostGroup=nil        => computeCluster c1 as failure domain
3. datacenter=dc1,computeCluster=c1,hostGroup=h1        => hostGroup h1 as failure domain
4. datacenter=dc1,computeCluster=nil,hostGroup=h1        => invalid input

Reply:
correct